Mathlib/Analysis/Analytic/OfScalars.lean

```diff
@@ -88,44 +158,142 @@ theorem ofScalars_norm_le (hn : n > 0) : ‖ofScalars E c n‖ ≤ ‖c n‖ :=
   simp [ofScalars_norm_eq_mul]
 
-/-- The radius of convergence of a scalar series is the inverse of the ratio of the norms of the
-coefficients. -/
-theorem ofScalars_radius_eq_inv_of_tendsto [NormOneClass E] {r : NNReal} (hr : r ≠ 0)
+private theorem tendsto_succ_norm_div_norm {r r' : ℝ≥0} (hr' : r' ≠ 0)
     (hc : Tendsto (fun n ↦ ‖c n.succ‖ / ‖c n‖) atTop (𝓝 r)) :
-      (ofScalars E c).radius = ENNReal.ofNNReal r⁻¹ := by
-  have hc' {r' : NNReal} (hr' : (r' : ℝ) ≠ 0) :
-    Tendsto (fun n ↦ ‖‖ofScalars E c (n + 1)‖ * r' ^ (n + 1)‖ /
-      ‖‖ofScalars E c n‖ * r' ^ n‖) atTop (𝓝 ↑(r' * r)) := by
-    simp_rw [norm_mul, norm_norm, ofScalars_norm, mul_div_mul_comm, ← norm_div, pow_succ,
-      mul_div_right_comm, div_self (pow_ne_zero _ hr'), one_mul, norm_div, NNReal.norm_eq]
-    exact mul_comm r' r ▸ Filter.Tendsto.mul hc tendsto_const_nhds
-  apply le_antisymm <;> refine ENNReal.le_of_forall_nnreal_lt (fun r' hr' ↦ ?_)
-  · rw [ENNReal.coe_le_coe, NNReal.le_inv_iff_mul_le hr]
-    have := FormalMultilinearSeries.summable_norm_mul_pow _ hr'
-    contrapose! this
-    have hrz : (r' : ℝ) ≠ 0 := by aesop
-    apply not_summable_of_ratio_test_tendsto_gt_one this
-    exact hc' (by aesop)
-  · rw [ENNReal.coe_lt_coe, NNReal.lt_inv_iff_mul_lt hr] at hr'
-    by_cases hrz : r' = 0
-    · simp [hrz]
-    · apply FormalMultilinearSeries.le_radius_of_summable_norm
-      refine summable_of_ratio_test_tendsto_lt_one hr' ?_ <| hc' (NNReal.coe_ne_zero.2 hrz)
-      refine (Tendsto.eventually_ne hc (NNReal.coe_ne_zero.2 hr)).mp (Eventually.of_forall ?_)
-      simp_rw [div_ne_zero_iff, ofScalars_norm, mul_ne_zero_iff]
-      aesop
+      Tendsto (fun n ↦ ‖‖c (n + 1)‖ * r' ^ (n + 1)‖ /
+        ‖‖c n‖ * r' ^ n‖) atTop (𝓝 ↑(r' * r)) := by
+  simp_rw [norm_mul, norm_norm, mul_div_mul_comm, ← norm_div, pow_succ,
+    mul_div_right_comm, div_self (pow_ne_zero _ (NNReal.coe_ne_zero.mpr hr')
+    ), one_mul, norm_div, NNReal.norm_eq]
+  exact mul_comm r' r ▸ hc.mul tendsto_const_nhds
+
+theorem ofScalars_radius_ge_inv_of_tendsto {r : ℝ≥0} (hr : r ≠ 0)
+    (hc : Tendsto (fun n ↦ ‖c n.succ‖ / ‖c n‖) atTop (𝓝 r)) :
+      (ofScalars E c).radius ≥ ofNNReal r⁻¹ := by
+  refine le_of_forall_nnreal_lt (fun r' hr' ↦ ?_)
+  rw [coe_lt_coe, NNReal.lt_inv_iff_mul_lt hr] at hr'
+  by_cases hrz : r' = 0
+  · simp [hrz]
+  · apply FormalMultilinearSeries.le_radius_of_summable_norm
+    refine Summable.of_norm_bounded_eventually (fun n ↦ ‖‖c n‖ * r' ^ n‖) ?_ ?_
+    · refine summable_of_ratio_test_tendsto_lt_one hr' ?_ ?_
+      · refine (hc.eventually_ne (NNReal.coe_ne_zero.mpr hr)).mp (Eventually.of_forall ?_)
+        aesop
+      · simp_rw [norm_norm]
+        exact tendsto_succ_norm_div_norm c hrz hc
+    · filter_upwards [eventually_cofinite_ne 0] with n hn
+      simp only [norm_mul, norm_norm, norm_pow, NNReal.norm_eq]
+      gcongr
+      exact ofScalars_norm_le E c n (Nat.pos_iff_ne_zero.mpr hn)
```

Comment:
We try not to indent the main branch, cf the "line of sight rule"
```suggestion
  apply FormalMultilinearSeries.le_radius_of_summable_norm
  refine Summable.of_norm_bounded_eventually (fun n ↦ ‖‖c n‖ * r' ^ n‖) ?_ ?_
  · refine summable_of_ratio_test_tendsto_lt_one hr' ?_ ?_
    · refine (hc.eventually_ne (NNReal.coe_ne_zero.mpr hr)).mp (Eventually.of_forall ?_)
      aesop
    · simp_rw [norm_norm]
      exact tendsto_succ_norm_div_norm c hrz hc
  · filter_upwards [eventually_cofinite_ne 0] with n hn
    simp only [norm_mul, norm_norm, norm_pow, NNReal.norm_eq]
    gcongr
    exact ofScalars_norm_le E c n (Nat.pos_iff_ne_zero.mpr hn)
```

Reply:
Is the main branch the "last open goal", so to speak? Because by_cases generates two branches of equal "importance", so I like to treat them equally 😆

Reply:
Yes. But the first branch is clearly a degenerate case here ;)